[SPARK-22372][core, yarn] Make cluster submission use SparkApplication.

vanzin opened the pull request:
The main goal of this change is to allow multiple cluster-mode
submissions from the same JVM, without having them end up with
mixed configuration. That is done by extending the SparkApplication
trait, and doing so was reasonably trivial for standalone and
mesos modes.

For YARN mode, there was a complication. YARN used a "SPARK_YARN_MODE"
system property to control behavior indirectly in a whole bunch of
places, mainly in the SparkHadoopUtil / YarnSparkHadoopUtil classes.
Most of the changes here are removing that.

Since we removed support for Hadoop 1.x, some methods that lived in
YarnSparkHadoopUtil can now live in SparkHadoopUtil. The remaining
methods don't need to be part of the class, and can be called directly
from the YarnSparkHadoopUtil object, so now there's a single
implementation of SparkHadoopUtil.

There were two places in the code that relied on  SPARK_YARN_MODE to
make decisions about YARN-specific functionality, and now explicitly check
the master from the configuration for that instead:

* fetching the external shuffle service port, which can come from the YARN
  configuration. 

* propagation of the authentication secret using Hadoop credentials. This also
  was cleaned up a little to not need so many methods in `SparkHadoopUtil`.

With those out of the way, actually changing the YARN client
to extend SparkApplication was easy.

Tested with existing unit tests, and also by running YARN apps
with auth and kerberos both on and off in a real cluster.

Commit:
[SPARK-22372][core, yarn] Make cluster submission use SparkApplication.

The main goal of this change is to allow multiple cluster-mode
submissions from the same JVM, without having them end up with
mixed configuration. That is done by extending the SparkApplication
trait, and doing so was reasonably trivial for standalone and
mesos modes.

For YARN mode, there was a complication. YARN used a "SPARK_YARN_MODE"
system property to control behavior indirectly in a whole bunch of
places, mainly in the SparkHadoopUtil / YarnSparkHadoopUtil classes.
Most of the changes here are removing that.

Since we removed support for Hadoop 1.x, some methods that lived in
YarnSparkHadoopUtil can now live in SparkHadoopUtil. The remaining
methods don't need to be part of the class, and can be called directly
from the YarnSparkHadoopUtil object, so now there's a single
implementation of SparkHadoopUtil.

One remaining use case was fetching the external shuffle
service port, which can come from the YARN configuration. That
is now done by checking the master used to submit the app,
instead of the system property.

The other use case was the propagation of the auth secret.
That was done by stashing the secret in the current user's
credentials in YARN mode. Instead, the secret is now propagated
using the config / env variable like for standalone and mesos.
That allowed a few methods in SparkHadoopUtil to go away.
There's still a little bit of code in SecurityManager that
is currently YARN-specific, but that uses the conf's master
to detect whether the app is a YARN app. This also has the
benefit of not stashing the secret in a shared location (the
current UGI), making sure different apps use different secrets.

With those out of the way, actually changing the YARN client
to extend SparkApplication was easy.

Tested with existing unit tests, and also by running YARN apps
with auth and kerberos both on and off in a real cluster.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83297/
Test FAILed.

vanzin commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83298/
Test FAILed.

vanzin commented:
@shaneknapp any idea what's going on?

```
chmod: cannot access `target/*': No such file or directory
```

vanzin commented:
retest this please

SparkQA commented:
**[Test build #83299 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83299/testReport)** for PR 19631 at commit [`c80554b`](https://github.com/apache/spark/commit/c80554bed52f96ada2e1c4b6a16f5f3e5c7e5317).

SparkQA commented:
**[Test build #83299 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83299/testReport)** for PR 19631 at commit [`c80554b`](https://github.com/apache/spark/commit/c80554bed52f96ada2e1c4b6a16f5f3e5c7e5317).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83299/
Test FAILed.

Commit:
Add mima excludes.

SparkHadoopUtil is not documented as a public class in Spark's API
docs, so just added new exclusions.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83302/
Test FAILed.

vanzin commented:
retest this please

SparkQA commented:
**[Test build #83305 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83305/testReport)** for PR 19631 at commit [`cee6be2`](https://github.com/apache/spark/commit/cee6be231310b89d35d5b419bd30153f29dd4cb9).

vanzin commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83349/
Test FAILed.

vanzin commented:
Sigh. retest this please

SparkQA commented:
**[Test build #83352 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83352/testReport)** for PR 19631 at commit [`cee6be2`](https://github.com/apache/spark/commit/cee6be231310b89d35d5b419bd30153f29dd4cb9).

SparkQA commented:
**[Test build #83352 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83352/testReport)** for PR 19631 at commit [`cee6be2`](https://github.com/apache/spark/commit/cee6be231310b89d35d5b419bd30153f29dd4cb9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83352/
Test PASSed.

shaneknapp commented:
That's an expected message from the build and has no impact on it. I'll
change the build config to suppress the message when I'm back from holiday
late next week.

On Nov 1, 2017 6:11 PM, "Marcelo Vanzin" <notifications@github.com> wrote:

@shaneknapp <https://github.com/shaneknapp> any idea what's going on?

chmod: cannot access `target/*': No such file or directory

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/apache/spark/pull/19631#issuecomment-341191716>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/ABiDrNcw4Zj27px49Fm53DRl7O4W1T2xks5syLRsgaJpZM4QOkme>
.

ArtRand commented on core/src/main/scala/org/apache/spark/SecurityManager.scala:
```diff
@@ -542,7 +496,55 @@ private[spark] class SecurityManager(
+      return
+    }
+
+    // In YARN, force creation of a new secret if this is client mode. This ensures each
```
Is there a reason this _has_ to be unique to YARN? Will this solve the problem (in Mesos currently) where when the Executors [bootstrap](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala#L193) they do so without security (unless you "bake" the secret and secret config into the container image)?

ArtRand commented on core/src/main/scala/org/apache/spark/SecurityManager.scala:
```diff
@@ -542,7 +496,55 @@ private[spark] class SecurityManager(
+      return
+    }
+
+    // In YARN, force creation of a new secret if this is client mode. This ensures each
```
Is there a reason this _has_ to be unique to YARN? Will this solve the problem (in Mesos currently) where when the Executors [bootstrap](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala#L193) they do so without security (unless you "[bake](https://github.com/apache/spark/blob/e1960c3d6f380b0dfbba6ee5d8ac6da4bc29a698/core/src/main/scala/org/apache/spark/SparkConf.scala#L482)" the secret and secret config into the container image)? Looks like propagating the envvar is only handled in the YARN case?

vanzin replied:
This behaves the same way as before for non-YARN. Standalone and Mesos have always used hardcoded secrets in the config to authenticate executors to driver and the driver to the master (in the case of standalone).

You can see the code I'm changing in this class, where for non-YARN it would throw an error if the secret was not set. If changing that behavior is desired for Mesos, then it should be done in a separate change.

vanzin replied:
BTW standalone at least propagates the secret using an env var, the issue is just that standalone, at least, needs the same secret everywhere, including the part where the driver authenticates with the master. Mesos just inherited that.

ArtRand replied:
Yes. I guess I'm wondering if now with your change will this work in all cases not just YARN? Perhaps obviously, I'm looking into changing this for Mesos.

vanzin replied:
I'm not changing the previous behavior for non-YARN. So my change shouldn't make it easier nor harder to make things work for other cluster managers.

Whether it will work depends on how the auth secret is used in those cases.

ArtRand replied:
Ok, thanks for the clarification.

ArtRand commented on core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:
```diff
@@ -365,22 +370,21 @@ object SparkSubmit extends CommandLineUtils with Logging {
 
     // This security manager will not need an auth secret, but set a dummy value in case
     // spark.authenticate is enabled, otherwise an exception is thrown.
```
this comment is no longer true?

ArtRand commented on core/src/main/scala/org/apache/spark/deploy/rest/RestSubmissionClient.scala:
```diff
@@ -398,9 +399,20 @@ private[spark] object RestSubmissionClient {
+    env.filterKeys { k =>
+      // SPARK_HOME is filtered out because it is usually wrong on the remote machine (SPARK-12345)
+      (k.startsWith("SPARK_") && k != "SPARK_ENV_LOADED" && k != "SPARK_HOME") ||
+        k.startsWith("MESOS_")
```
Will this may break Mesos when using [the mesos bundle](https://github.com/apache/spark/blob/master/conf/spark-env.sh.template#L33)?

vanzin replied:
I don't know, but this should be the exact same behavior as before, no? I didn't really change this code.

ArtRand replied:
Yes, I apologize you're correct. I think this is actually to filter out things like `MESOS_EXECUTOR_ID` and `MESOS_FRAMEWORK_ID`

Commit:
Remove stale comment.

SparkQA commented:
**[Test build #83555 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83555/testReport)** for PR 19631 at commit [`cb6c8ff`](https://github.com/apache/spark/commit/cb6c8ffe9eb04e05b8cd982e123ae962bd2e7373).

SparkQA commented:
**[Test build #83555 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83555/testReport)** for PR 19631 at commit [`cb6c8ff`](https://github.com/apache/spark/commit/cb6c8ffe9eb04e05b8cd982e123ae962bd2e7373).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83555/
Test PASSed.

vanzin commented:
@tgravescs @jerryshao

jerryshao commented on core/src/main/scala/org/apache/spark/SecurityManager.scala:
```diff
@@ -551,13 +553,10 @@ private[spark] class SecurityManager(
-  val SPARK_AUTH_CONF: String = "spark.authenticate"
-  val SPARK_AUTH_SECRET_CONF: String = "spark.authenticate.secret"
+  val SPARK_AUTH_CONF = NETWORK_AUTH_ENABLED.key
+  val SPARK_AUTH_SECRET_CONF = "spark.authenticate.secret"
```
I think we can also make this as a `ConfigEntry`.

vanzin replied:
In a separate change maybe. There are too many references to this right now, it would be really noisy here.

jerryshao commented on core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:
```diff
@@ -92,6 +92,11 @@ object SparkSubmit extends CommandLineUtils with Logging {
+  // Following constants are visible for testing.
+  private[deploy] val YARN_SUBMIT_CLASS = "org.apache.spark.deploy.yarn.YarnClusterApplication"
+  private[deploy] val REST_SUBMIT_CLASS = classOf[RestSubmissionClientApp].getName()
+  private[deploy] val STANDALONE_SUBMIT_CLASS = classOf[ClientApp].getName()
```
Maybe we can rename these variables to `XXX_CLUSTER_SUBMIT_CLASS`.

vanzin replied:
Sure.

jerryshao commented on core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:
```diff
@@ -216,7 +216,9 @@ private[spark] object CoarseGrainedExecutorBackend extends Logging {
         logInfo("Will periodically update credentials from: " +
           driverConf.get("spark.yarn.credentials.file"))
-        SparkHadoopUtil.get.startCredentialUpdater(driverConf)
+        Utils.classForName("org.apache.spark.deploy.yarn.YarnSparkHadoopUtil")
```
What's the purpose of changing to this way?

vanzin replied:
`YarnSparkHadoopUtil` does not extend `SparkHadoopUtil` anymore (it doesn't even exist as a class), so I'm using this to invoke the method on the YARN object.

I plan to clean this up once #19272 is in, since that's a cleaner way to do this token renewal thing.

jerryshao replied:
I see, thanks for explanation, this kind of reflection seems not so elegant.

jerryshao replied:
#19272 is already in, do you plan to update it here in this PR or in a separate PR?

vanzin replied:
I plan to do it in a separate PR (also to make YARN and Mesos share more code in that area).

jerryshao reviewed (commented):
This PR seems also involve many different changes in many files, needs to understand again.

Commit:
Rename constants.

SparkQA commented:
**[Test build #83858 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83858/testReport)** for PR 19631 at commit [`593eb6b`](https://github.com/apache/spark/commit/593eb6b67c871706d7f4b418dfa6fb111ec06fde).

SparkQA commented:
**[Test build #83858 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83858/testReport)** for PR 19631 at commit [`593eb6b`](https://github.com/apache/spark/commit/593eb6b67c871706d7f4b418dfa6fb111ec06fde).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83858/
Test PASSed.

jerryshao commented on resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -745,15 +739,20 @@ private[spark] class Client(
-      Option(amKeytabFileName).foreach { k => props.setProperty(KEYTAB.key, k) }
+      Option(amKeytabFileName).foreach { k =>
+        // Do not propagate the app's secret using the config file.
+        if (k != SecurityManager.SPARK_AUTH_SECRET_CONF) {
```
Is it necessary to add a check here? I'm not sure how this could happen.

vanzin replied:
Oh, I think this is the wrong place for the check.

jerryshao reviewed (commented):
Just did another round of review and leave some questions, LGTM overall.

One more question, does this PR implies that supporting to launch multiple SparkSubmit in one JVM will only work in cluster mode?

vanzin commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -412,8 +412,6 @@ class SparkContext(config: SparkConf) extends Logging {
       }
     }
 
-    if (master == "yarn" && deployMode == "client") System.setProperty("SPARK_YARN_MODE", "true")
```
This change is removing all references to `SPARK_YARN_MODE`.

Commit:
Correctly filter out SPARK_AUTH_SECRET_CONF.

Commit:
Cleanup SPARK_YARN_MODE in tests.

SparkQA commented:
**[Test build #83875 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83875/testReport)** for PR 19631 at commit [`121bcf8`](https://github.com/apache/spark/commit/121bcf8a2758858cad8e88e7fb7d78566494765b).

SparkQA commented:
**[Test build #83876 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83876/testReport)** for PR 19631 at commit [`08f47ca`](https://github.com/apache/spark/commit/08f47ca3fb54315c537b1134e31e0a1a912c285e).

SparkQA commented:
**[Test build #83875 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83875/testReport)** for PR 19631 at commit [`121bcf8`](https://github.com/apache/spark/commit/121bcf8a2758858cad8e88e7fb7d78566494765b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83875/
Test PASSed.

SparkQA commented:
**[Test build #83876 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83876/testReport)** for PR 19631 at commit [`08f47ca`](https://github.com/apache/spark/commit/08f47ca3fb54315c537b1134e31e0a1a912c285e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ClientSuite extends SparkFunSuite with Matchers `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83876/
Test PASSed.

jerryshao commented:
Did another round of review, LGTM overall. @tgravescs do you any comment?

tgravescs commented:
Sorry on vacation til Monday. I would like to look at this though since it says changing secret passing to env variable.

Commit:
Propagate secret to executors from AM to avoid writing it to disk.

SparkQA commented:
**[Test build #83946 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83946/testReport)** for PR 19631 at commit [`2129ccb`](https://github.com/apache/spark/commit/2129ccb13759fbc7784ba31e7c0e6aa08904cb42).

SparkQA commented:
**[Test build #83946 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83946/testReport)** for PR 19631 at commit [`2129ccb`](https://github.com/apache/spark/commit/2129ccb13759fbc7784ba31e7c0e6aa08904cb42).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83946/
Test FAILed.

Commit:
Unbreak test.

SparkQA commented:
**[Test build #83951 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83951/testReport)** for PR 19631 at commit [`86f0bf8`](https://github.com/apache/spark/commit/86f0bf8dcdd4a3dfc43f0e1f815b348f3477b250).

SparkQA commented:
**[Test build #83951 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83951/testReport)** for PR 19631 at commit [`86f0bf8`](https://github.com/apache/spark/commit/86f0bf8dcdd4a3dfc43f0e1f815b348f3477b250).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83951/
Test PASSed.

tgravescs commented:
fyi - in progress of reviewing should be done in an hour or so.

tgravescs reviewed (commented):
My only other concern here is that an environment variable I think opens up more possibility for leaking the secret as well. For instance if someone is starting a server that is getting hit by other users that server could be started with the same env and then inadvertently expose the secret to other users.  For this, you could argue they should handle but I think its just something that could happen more easily then if the secret is in the hadoop credentials.

ArtRand commented on core/src/main/scala/org/apache/spark/SecurityManager.scala:
```diff
@@ -542,7 +496,54 @@ private[spark] class SecurityManager(
-  def getSecretKey(): String = secretKey
+  def getSecretKey(): String = {
+    if (isAuthenticationEnabled) {
+      Option(sparkConf.getenv(ENV_AUTH_SECRET))
```
FWIW, in Mesos, we are planning on using the [Secrets primitives](https://github.com/apache/spark/blob/b3f9dbf48ec0938ff5c98833bb6b6855c620ef57/resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/config.scala#L26) to distribute `ENV_AUTH_SECRET`. This way Mesos and YARN can both use the same secret-generation code and only differ in the distribution of the secret. `SPARK_AUTH_SECRET_CONF` is already somewhat awkward because it has to be in the config.

vanzin replied:
I can add a different, internal config for this is re-using `SPARK_AUTH_SECRET_CONF` is confusing. But I'm not too concerned about exposing this to the user code running the application; they can just as easily get that info from the UGI currently. Spark already redacts this kind of information when writing it to things like the event log, which would be one place where it might leak out.

vanzin replied:
I changed things and now they work pretty much as before. It would be good to separate secret generation from distribution, but I'd rather do that separately.

tgravescs replied:
I agree SPARK_AUTH_SECRET_CONF is awkward and not really secure, when I initially did this , this is what was requested by other committers since standalone and mesos needed more security work around it anyway.

I don't follow how the MesosSecretConfig is going to be used fully.  Are these just regular spark configs passed around or distributed through mesos somehow?

Commit:
Merge branch 'master' into SPARK-22372

vanzin commented on resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -58,18 +58,14 @@ import org.apache.spark.util.{CallerContext, Utils}
-
   private val yarnClient = YarnClient.createYarnClient
-  private val yarnConf = new YarnConfiguration(hadoopConf)
+  private val hadoopConf = new YarnConfiguration(SparkHadoopUtil.newConfiguration(sparkConf))
```
Both are effectively the same; I chose too keep the name used more often to minimize changes.

vanzin commented:
> For instance if someone is starting a server that is getting hit by other users that server could be started with the same env and then inadvertently expose the secret to other users.

If I understand what you're saying correctly, that should be considered a security issue in that server application regardless of this change. The server should not be exposing its environment to unprivileged users.

That being said, it seems Spark's own thrift server does that. If the following works in a spark-shell, it probably would do the same through the STS:

```
scala> spark.sql("set spark.sql.columnNameOfCorruptRecord=${env:SCALA_HOME}").show()
+--------------------+------------------+
|                 key|             value|
+--------------------+------------------+
|spark.sql.columnN...|/apps/scala-2.11.7|
+--------------------+------------------+
```

So this change would expose that secret to users also in YARN mode (it's already exposed in Standalone and Mesos currently, because it's in the config).

Let me think a little about this. I prefer the environment to the credentials approach because the latter are written to disk, but at the same time, that's less problematic than exposing the secret to users in the STS.

Commit:
Feedback.

Commit:
Go back to propagating secret using credentials.

SparkQA commented:
**[Test build #84363 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/84363/testReport)** for PR 19631 at commit [`c752453`](https://github.com/apache/spark/commit/c752453b2a379f301d52692bfc639bb631520069).

SparkQA commented:
**[Test build #84363 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/84363/testReport)** for PR 19631 at commit [`c752453`](https://github.com/apache/spark/commit/c752453b2a379f301d52692bfc639bb631520069).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/84363/
Test PASSed.

jerryshao commented:
LGTM.

tgravescs commented on project/MimaExcludes.scala:
```diff
@@ -36,6 +36,12 @@ object MimaExcludes {
   // Exclude rules for 2.3.x
   lazy val v23excludes = v22excludes ++ Seq(
+    // SPARK-22372: Make cluster submission use SparkApplication.
+    ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.deploy.SparkHadoopUtil.getSecretKeyFromUserCredentials"),
```
This is only going into spark 2.3 ?

vanzin replied:
The change is 2.3 only.

I've always questioned why `SparkHadoopUtil` is public in the first place. I'm not that worried about removing the functionality in this method because Spark apps don't really need to depend on it; it's easy to call the Hadoop API directly, and the API did nothing before except in YARN mode.

tgravescs commented:
> If I understand what you're saying correctly, that should be considered a security issue in that server application regardless of this change. The server should not be exposing its environment to unprivileged users.

Yes this is what i was saying and agree one could argue its their fault its more of us being nice because we are not adding it in and they might not realize its something they shouldn't expose.  I'm not sure if there was anything that sensitive in there before but still good practice to not expose either way.


> Let me think a little about this. I prefer the environment to the credentials approach because the 
>  latter are written to disk, but at the same time, that's less problematic than exposing the secret to users in the STS.
> 

I agree written to disk isn't ideal either but its only local disk and not HDFS so user would have to compromise the box vs someone simply accidentally allowing access to env/conf via a ui or console.

reviewing changes now.

tgravescs commented:
+1

vanzin commented:
Merging to master.
